### DISCUSSION TITLE
YALMIP and nullspace

### DISCUSSION THREAD

**Discussion by kirkrudolph** [yalmip/YALMIP#1291] [Problems using YALMIP] [answered]
Hi Johan,

Back again with a different optimization problem. I believe this one should be easier. Here I have 31 decision variables. The trouble I'm having is the cost function involves the null space of the decision variables and I'm getting an error trying to use the `null` function on the `sdpvar` types.

I didn't see `null.m` in the `YALMIP/@sdpvar` folder so I'm assuming it simply doesn't have an overload?

## Problem Pseudo Code

``` matlab
% Decision Variables
r = sdpvar(31,1);

% Constant Weights
J = rand(21,1);
```

``` matlab
% Cost Function
Q = foo( r );                                 % Sparse [19 21] matrix
Objective = null( Q )'*diag( J )*null( Q );   % [2 21] x [21 21] x [21 2]
```

## Troubleshooting 

I verified Q is what I'd expect it to be:
```
>> Q
Linear matrix variable 19x21 (full, real, 31 variables)
Coeffiecient range: 1 to 1
```

## Error Message
```
Check for missing argument or incorrect argument data type in call to function 'isfinite'.

Error in rat (line 57)
    tol = 1.e-6*norm(X(isfinite(X)),1);

Error in rref (line 26)
[num, den] = rat(A);

Error in null (line 50)
    [R,pivcol] = rref(A);

Error in YALMIP_Optimize (line 17)
null( Q, "rational" );
 ```

I tried using some other functions too
```
>> null(Q)
Error using svd
First input must be single or double.

Error in null (line 65)
    [~,S,V] = svd(A,0);
```

Thanks,
Kirk Rudolph

**Comment by johanlofberg** [comment, marked as answer]
Not sure what you expect to be returned from a null operator acting on a matrix involving decision variables

Just randomly, if

Q = [r1 r2 0;r2 r1 0] 

then for any point in the solver iteration process where r1=r2 the result would be a 3x2 vector, and when r1 and r2 are different it is a 3x1 vector. There is no way to work with such an operator in a standard optimization framework or represent in YALMIP, as everything has defined sizes when the model is set up

**Reply by johanlofberg** [reply]
The only way to be able to attack it would be a black-box operator in a general nonlinear solver

https://yalmip.github.io/command/blackbox/

Leads to very slow and inefficient code (as YALMIP is not built for this) and probably poor performance in the solver as no derivatives are supplied.

I also believe you are missing a trace

`objective = blackbox(r,@(r)trace((null(foo(r))'*diag(J)*null(foo(r)))))`

**Reply by kirkrudolph** [reply]
Hmm, yeah I never thought about how certain selections of the decision variables could change the size of the null space. For my problem, the result of the null operator should always be a fixed size of [21 2].

**Reply by johanlofberg** [reply]
What you have to do then is to figure out if you can write that fixed null-space as a function of r, from the basis of Q (getbase(Q))

**Reply by kirkrudolph** [reply]
That makes sense. I don't believe it'll be too hard to make a function that computes the [21 2] matrix as a function of r. 

The null operator is the generic way that works well for any "structure" of Q with predetermined values of r but this optimization flips everything around from work I've done in the past. It now requires a fixed "structure" of Q and varying values of r. I can see how using null would be problematic when the values are changing.

Thanks again for your help!
